# Table of contents <br>
 - [Introduction](#introduction)
 - [Analysis workflow](#analysis_workflow)
 - [How to use this tutorial](#how_to_use_this_tutorial)
 - [Broad peakcalling](#broad_peakcalling)

## Introduction <br>
This tutorial walks step-by-step tutorial of analysis pipeline for ChIP-seq/CUT&RUN. In my experience, I found you can generally use the same analysis workflow for the two types of experiment, but there are studies proposing tailored CUT&RUN analysis tools such as [SEACR](https://epigeneticsandchromatin.biomedcentral.com/articles/10.1186/s13072-019-0287-4), you are welcome to experimening orthogonal approaches and becnchmark their performance. 

## Analysis workflow
![GitHub Logo]()
## How to use this tutorial
* ### set up your working directory
* ### Trim fastq files for sequencing adaptors
  * #### If you don't know the sequencing adaptor 
* ### modify the job submission file (wrapper file)
* ### change directory to the folder that contains your job file and job submission file, type in ``` chmod +x * ```
## Broad peak calling



